egs/librispeech/ASR/pruned_transducer_stateless3/onnx_check_all_in_one.py

```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 Xiaomi Corporation (Author: Yunus Emre Ozkose)
+#
+# See ../../../../LICENSE for clarification regarding multiple authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This script checks that exported onnx models produce the same output
+with the given torchscript model for the same input.
+"""
+
+import os
+import argparse
+import logging
+
+import onnxruntime as ort
+import torch
+
+import onnx
+import onnxruntime
+import onnx_graphsurgeon as gs
+
+ort.set_default_logger_severity(3)
+
+
+def get_parser():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+
+    parser.add_argument(
+        "--jit-filename",
+        required=True,
+        type=str,
+        help="Path to the torchscript model",
+    )
+
+    parser.add_argument(
+        "--onnx-encoder-filename",
+        required=True,
+        type=str,
+        help="Path to the onnx encoder model",
+    )
+
+    parser.add_argument(
+        "--onnx-decoder-filename",
+        required=True,
+        type=str,
+        help="Path to the onnx decoder model",
+    )
+
+    parser.add_argument(
+        "--onnx-joiner-filename",
+        required=True,
+        type=str,
+        help="Path to the onnx joiner model",
+    )
+
+    parser.add_argument(
+        "--onnx-all-in-one-filename",
+        required=True,
+        type=str,
+        help="Path to the onnx all in one model",
+    )
+
+    return parser
+
+
+def test_encoder(
```

Comment:
Could you use
```python
from onnx_check import test_encoder
```
at the beginning and remove this function from the current file?

Reply:
Since prefixes are added to node names during [exporting](https://github.com/EmreOzkose/icefall/blob/18a6b8afad1cb29cf03eec06afed77cbd820baed/egs/librispeech/ASR/pruned_transducer_stateless3/export.py#L526), input and output names are changed. Hence in/out names in [test_encoder](https://github.com/EmreOzkose/icefall/blob/18a6b8afad1cb29cf03eec06afed77cbd820baed/egs/librispeech/ASR/pruned_transducer_stateless3/onnx_check_all_in_one.py#L64) are different. I tried to fix names with this function:

```
def fix_names(onnx_model):
    for i in range(len(onnx_model.graph.node)):
        for j in range(len(onnx_model.graph.node[i].input)):
            onnx_model.graph.node[i].input[j] = onnx_model.graph.node[i].input[j].replace('encoder/', '')

        for j in range(len(onnx_model.graph.node[i].output)):
            onnx_model.graph.node[i].output[j] = onnx_model.graph.node[i].output[j].replace('encoder/', '')

        onnx_model.graph.node[i].name = onnx_model.graph.node[i].name.replace('encoder/', '')

    for i in range(len(onnx_model.graph.input)):
        onnx_model.graph.input[i].name = onnx_model.graph.input[i].name.replace('encoder/', '')

    for i in range(len(onnx_model.graph.output)):
        onnx_model.graph.output[i].name = onnx_model.graph.output[i].name.replace('encoder/', '')

    return onnx_model

def extract_encoder(onnx_model: onnx.ModelProto):
    encoder_ = extract_sub_model(onnx_model, ["encoder/x", "encoder/x_lens"], ["encoder/encoder_out", "encoder/encoder_out_lens"], False)
    onnx_model = fix_names(onnx_model)    # only 1 line is added to fix names
    onnx.save(encoder_, "tmp_encoder.onnx")
    onnx.checker.check_model(encoder_)
    sess = onnxruntime.InferenceSession("tmp_encoder.onnx")
    os.remove("tmp_encoder.onnx")
    return sess
```

, but I got this error

```
2022-08-03 15:18:41,530 INFO [onnx_check.py:249] Test encoder
Traceback (most recent call last):
  File "./pruned_transducer_stateless3/onnx_check.py", line 277, in <module>
    main()
  File "/home/yunusemre.ozkose/anaconda3/envs/k2_sherpa/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "./pruned_transducer_stateless3/onnx_check.py", line 250, in main
    encoder_session = extract_encoder(onnx_model)
  File "./pruned_transducer_stateless3/onnx_check.py", line 232, in extract_encoder
    onnx.checker.check_model(encoder_)
  File "/home/yunusemre.ozkose/anaconda3/envs/k2_sherpa/lib/python3.8/site-packages/onnx/checker.py", line 106, in check_model
    C.check_model(protobuf_string)
onnx.onnx_cpp2py_export.checker.ValidationError: Nodes in a graph must be topologically sorted, however input 'encoder_embed.out_norm.eps' of node:
name: Exp_39 OpType: Exp
```

Reply:
Ah, could you remove the check for input/output names?

Reply:
Do you mean [these checks](https://github.com/EmreOzkose/icefall/blob/40126355ecc88424ab815a2217b58ee160bb1ad6/egs/librispeech/ASR/pruned_transducer_stateless3/onnx_check_all_in_one.py#L64)?

Reply:
Yes, you also need to replace
https://github.com/k2-fsa/icefall/blob/e538232485db8999ee2a5308685cc61e3c4f79b0/egs/librispeech/ASR/pruned_transducer_stateless3/onnx_check.py#L87
with
```python3
encoder_inputs[0].name: x.numpy(),
```